src/ops_complex.rs

```diff
@@ -551,14 +551,17 @@ fn generate_permutation_matrix_raw(
 }
 
 pub fn get_generate_permutation_matrix_temp_storage_bytes(num_cells: usize) -> CudaResult<usize> {
-    let end_bit: i32 = (usize::BITS - num_cells.leading_zeros()) as i32;
+    let cell_size = mem::size_of::<u32>();
+    let num_bytes = num_cells * cell_size;
     let sort_pairs_tsb =
-        get_sort_pairs_temp_storage_bytes::<u32, u32>(false, num_cells as u32, 0, end_bit)?;
+        get_sort_pairs_temp_storage_bytes::<u32, u32>(false, num_cells as u32, 0, 32)?;
+    assert!(sort_pairs_tsb <= 4 * num_bytes);
     let encode_tsb = get_encode_temp_storage_bytes::<u32>(num_cells as i32)?;
+    assert!(encode_tsb <= 2 * num_bytes - cell_size);
     let scan_tsb =
         get_scan_temp_storage_bytes::<u32>(ScanOperation::Sum, false, false, num_cells as i32)?;
-    let cub_tsb = max([sort_pairs_tsb, encode_tsb, scan_tsb]).unwrap();
-    Ok((7 * num_cells + 1) * 4 + cub_tsb)
+    assert!(scan_tsb <= 2 * num_bytes);
+    Ok(5 * num_bytes)
```

Comment:
because of https://github.com/matter-labs/era-boojum-cuda/pull/24/files#diff-edcb5a62c098440045351dfd26c3d8440b57f5c7d64bedbb3504cec7e0bc36f4R607, should this be `5 * num_bytes + cell_size` just to be safe?

Reply:
actually that would force temp_storage to be a wasteful allocation. Maybe you should accept `num_runs_out` as a standalone value instead of slicing it from temp_storage. You'd have to change the API and shivini usage though :(

Reply:
> because of https://github.com/matter-labs/era-boojum-cuda/pull/24/files#diff-edcb5a62c098440045351dfd26c3d8440b57f5c7d64bedbb3504cec7e0bc36f4R607, should this be `5 * num_bytes + cell_size` just to be safe?

this is the assert for scan, the assert for encode is 3 lines above and includes '- cell_size'

Reply:
Gotcha. Part of my (mistaken) concern was a suspicion that `generate_permutation_matrix` never "resets" the splitting on temp_storage, because i missed https://github.com/matter-labs/era-boojum-cuda/pull/24/files#r1466996495.